active/PagerControl/PagerControl.md

```diff
@@ -0,0 +1,356 @@
+Hopefully we'll be able to copy it mostly verbatim.
+So the second audience is everyone that reads there to learn how
+and why to use this API. -->
+
+## Background
+<!-- Use this section to provide background context for the new API(s) 
+in this spec. -->
+
+<!-- This section and the appendix are the only sections that likely
+do not get copied to docs.microsoft.com; they're just an aid to reading this spec. -->
+
+<!-- If you're modifying an existing API, included a link here to the
+existing page(s) -->
+
+<!-- For example, this section is a place to explain why you're adding this API rather than
+modifying an existing API. -->
+
+<!-- For example, this is a place to provide a brief explanation of some dependent
+area, just explanation enough to understand this new API, rather than telling
+the reader "go read 100 pages of background information posted at ...". -->
+
+The absence of a standard UI pager control, last seen as DataPager in Silverlight, has been a pain point in both WPF and UWP that has forced a variety of unfavorable workarounds for developers. Inclusion of this control in UWP would resolve an ecosystem gap and enable developers to efficiently deliver UI paging experiences in their applications. Through XAML Islands, it may also provide an opportunity to benefit WPF developers working with DataGrid or ListView. The scope of this proposal is not to provide a data paging solution at this time, but to start with UI paging as a foundation on top of which data paging may be later integrated.
+
+
+## Description
+<!-- Use this section to provide a brief description of the feature.
+For an example, see the introduction to the PasswordBox control 
+(http://docs.microsoft.com/windows/uwp/design/controls-and-patterns/password-box). -->
+
+Pager control is a customizable pager UI widget with generic events that can be configured to navigate pages for several view controls.
+
+**Important APIs:** [PagerControl class](https://docs.microsoft.com/en-us/uwp/api/microsoft.ui.xaml.controls.pagercontrol)
+
+Pager control can be configured to use a ComboBox, NumberBox, or numeric Button panel as the core pager control.
+
+## Is this the right control? 
+
+Use a **PagerControl** to create an accessible navigation interface for views that have multiple pages to display. 
+
+## Examples
+<!-- Use this section to explain the features of the API, showing
+example code with each description. The general format is: 
+  feature explanation,
+  example code
+  feature explanation,
+  example code
+  etc.-->
+  
+<!-- Code samples should be in C# and/or C++/WinRT -->
+
+<!-- As an example of this section, see the Examples section for the PasswordBox control 
+(https://docs.microsoft.com/windows/uwp/design/controls-and-patterns/password-box#examples). -->
+
+A pager control can have several configurations, including these notable ones. 
+
+### ComboBox Pager Control
+
+![](images/pager-control-combobox.png)
```

Comment:
Why is there a scroll bar in this picture? When I use the PagerControl, doesn't that make the ScrollBar not happen?

Reply:
I think the idea is that we limit the number of item to present the user at once, for example only 20 results from a search. Currently either you would need to implement a pager control yourself or potentially scroll through hundred or thousand of items. 

With the pager control, you would be able to just show `n` items reducing the performance overhead of a search (you would only need to load the first `n` items, which could have an image for example, where loading all at once would be too costly).

Reply:
I see. I think my expectations are misaligned with the goals for this control, and it's flavored my comments. For large content I think the user experience should be scrolling (panning) or paginated, but not both.

Reply:
> I think the idea is that we limit the number of item to present the user at once, for example only 20 results from a search. Currently either you would need to implement a pager control yourself or potentially scroll through hundred or thousand of items.
> 
> With the pager control, you would be able to just show `n` items reducing the performance overhead of a search (you would only need to load the first `n` items, which could have an image for example, where loading all at once would be too costly).

In many forums (which make heavy use of pagination) you can sometimes find an option for how many items to view on each page.  Increase the number of items, decreases the number of pages.

This control is detached from the view with the contents in - so this is probably something outside of the scope of this control, but there should be as much helper code available to determine the page count, from the item count in the list.

Reply:
But wouldn't a purely paginated approach result in the number of pages essentially changing every time the pager control gets resized? So basically, the page that is currently displayed would need to change every time the app gets resized and update the number of pages as a result of that.

If the PagerControl would be agnostic of the items, essentially like the ItemsRepeater and its layouts, this would actually would be a very interesting control, though it might be confusing for users.

Reply:
> In many forums (which make heavy use of pagination) you can sometimes find an option for how many items to view on each page. Increase the number of items, decreases the number of pages.

> This control is detached from the view with the contents in - so this is probably something outside of the scope of this control, but there should be as much helper code available to determine the page count, from the item count in the list.

Unless then number of items per page is variable, essentially the number of pages would be given by: 
`NumberOfPages = roundUp(Items/ItemsPerPage)`

I don't think that there would be much a helper could do, since the developer would need to provide a way to determine the number of items per page, and if the developer already is there, the helper wouldn't help much more I think.

Reply:
It does change the number of pages as you resize the control, but I think that's a good thing, like the Read Mode in Word. Actually that's a good prototype we could look at for the experience.

Having both scrolling and pagination at the same time is indeed common on web pages (like the Issues list on github.com), but I've always considered that to be a UX sacrifice made necessary because it's easier to implement.

Reply:
What about the following scenario: You have the 6th page of 10 open and resize the app (and thus the pager control) to show more items. How would this reflect in the index of the current page and it's content? After all, if we start at page 1 with e.g. 10 items, the 6th page the way it was previously presented could potentially now be out of date and not present the items like it should be presented. Should the items on screen change every time the user resizes the app? Would this be expected if some items suddenly disappear while others might suddenly be appear onscreen?

Reply:
I have actually prototyped this before. I found the app but can't find the source unfortunately. Actually I prototyped it to prove it was a bad idea, but played with it and liked it :)

In that prototype, which was a big list, when you resize the control it updates the number of pages, and updates the current page as necessary to follow the selected item.

Reply:
> I have actually prototyped this before. I found the app but can't find the source unfortunately. Actually I prototyped it to prove it was a bad idea, but played with it and liked it :)

If you find the source or can share the app, I would love to try that out!

At the end of the day, it's the developers choice to allow scrolling or not inside the pager control content. To avoid any misunderstanding, we could update the screenshot to now shown a scrollbar.

Reply:
I agree I can update the screenshot to not include the scroll bar for simplicity. @chingucoding are you suggesting that list view and other layout views should have the ability to have both scrolling and pagination? One thing we are considering is adding a property to the layout views to choose between scrolling or paging but the developer would have to choose one or the other.

Reply:
That's an interesting idea. How would those control behave when the viewport gets smaller or larger? Either you allow scroll or you would need to change the number of items per page, and as a result of that, the current page number.

I think the bigger question here would be how the PagerControl should work. Would you give a control with huge height/width and the PagerControl splits it up (similar to ScrollViewer which takes care of scrolling and the scrollbar) or would the developer need to take care of paginating it's data?
To provide the first case, we would need to find an answer to the first question in this comment.
The second case would be easier for us to define and develop, but would have the disadvantage of being more difficult to use for the developer.

Reply:
When the size changes and we switch pages, we could possibly ensure that the current anchor (some item in the view) still remains in view after the resize even if the page numbers change. Giving a single number of items per page is much simpler but that means avoiding the scrollbar is harder. As already discussed above if we have variable sized items each page might have different number of items and we would need to measure all items ahead to know which items fall in a specific page. We could expose a list of ranges if we wanted that would tell the pager what items fall in each page, but that sounds hard to calculate without breaking virtualization or knowing more about the data/item sizes upfront.